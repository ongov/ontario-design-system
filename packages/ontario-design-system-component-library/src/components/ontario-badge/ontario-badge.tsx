import { Component, Prop, Element, h, Watch, AttachInternals } from '@stencil/core';

import { BadgeColour, BadgeColours, BadgeColourToClass } from './ontario-badge.types';
import { ConsoleMessageClass } from '../../utils/console-message/console-message';
import { validateValueAgainstArray } from '../../utils/validation/validation-functions';

@Component({
	tag: 'ontario-badge',
	styleUrl: 'ontario-badge.scss',
	shadow: true,
})
export class OntarioBadge {
	@Element() host: HTMLElement;
	@AttachInternals() internals: ElementInternals;

	/**
	 * The colour of the badge.
	 */
	@Prop({ mutable: true }) colour: BadgeColour = 'teal';

	/**
	 * The label for the badge.
	 *
	 * Offical guidance is to keep the label length within 15 characters.
	 */
	@Prop() label: string;

	/**
	 * An aria label for screen readers.
	 *
	 * Used to provide more context to screen readers if necessary.
	 *
	 * This property is optional.
	 *
	 * @example
	 * <ontario-badge aria-label-text="This training is currently in progress.">In progress</ontario-badge>
	 */
	@Prop({ mutable: true }) ariaLabelText?: string;

	/**
	 * Watch for changes in the `label` prop for validation purposes.
	 *
	 * If no `label` value or host.textContent is provided, a warning message will be printed.
	 */
	@Watch('label')
	validateLabel() {
		if (!this.getBadgeLabel()) {
			const message = new ConsoleMessageClass();
			message
				.addDesignSystemTag()
				.addMonospaceText(' a label ')
				.addRegularText('for')
				.addMonospaceText(' <ontario-badge> ')
				.addRegularText('was not provided.')
				.printMessage();
		}
	}

	/**
	 * Watch for changes to the `colour` prop.
	 *
	 * If no `colour` is passed, or if the `colour` does not match the `colour` options, a default value of 'teal' will be applied.
	 *
	 * @returns {BadgeColour}
	 */
	@Watch('colour')
	validateColour(): BadgeColour {
		if (this.colour) {
			if (validateValueAgainstArray(this.colour, BadgeColours)) {
				return this.colour;
			} else {
				const message = new ConsoleMessageClass();
				message
					.addDesignSystemTag()
					.addMonospaceText(` colour ${this.colour} `)
					.addRegularText('for')
					.addMonospaceText(' <ontario-badge> ')
					.addRegularText('does not match one of the valid colour types. A default colour of `teal` will be applied.')
					.printMessage();
			}
		}

		return this.setBadgeColour('teal');
	}

	/**
	 * Programatically set and return the colour prop.
	 *
	 * @prop {BadgeColour} colour
	 *
	 * @returns {BadgeColour}
	 */
	setBadgeColour(colour: BadgeColour): BadgeColour {
		this.colour = colour;
		return this.colour;
	}

	/**
	 * Returns badge label depending on if the label prop was set or
	 * if the host.textContent (text in-between the opening and closing tag) was set.
	 *
	 * The label prop takes priority.
	 *
	 * @returns {string | null}
	 */
	getBadgeLabel(): string | null {
		if (this.label) return this.label;

		// Ensure host and its textContent are defined and accessible (safe for SSR environments)
		if (typeof this.host?.textContent === 'string') {
			return this.host.textContent;
		}

		return null;
	}

	componentWillLoad() {
		this.validateColour();
		this.validateLabel();
	}

	render() {
		return (
			<span class={`ontario-badge ${BadgeColourToClass[this.colour]}`} aria-label={this.ariaLabelText}>
				{this.getBadgeLabel()}
			</span>
		);
	}
}
