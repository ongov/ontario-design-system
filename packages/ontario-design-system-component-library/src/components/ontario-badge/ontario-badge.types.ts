import { retrieveEnumKeys } from '../../utils/helper/utils';

export enum BadgeColourToClass {
	teal = 'ontario-badge--default-heavy',
	lightTeal = 'ontario-badge--default-light',
	black = 'ontario-badge--neutral-heavy',
	grey = 'ontario-badge--neutral-light',
	red = 'ontario-badge--alert-heavy',
	yellow = 'ontario-badge--warning-heavy',
	green = 'ontario-badge--success-heavy',
	white = 'ontario-badge--white',
	darkGrey = 'ontario-badge--grey',
}

export const BadgeColours = retrieveEnumKeys(BadgeColourToClass);

export type BadgeColour = keyof typeof BadgeColourToClass;
