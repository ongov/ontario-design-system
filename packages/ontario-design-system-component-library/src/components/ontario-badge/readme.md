import { OntarioBadge } from '@ongov/ontario-design-system-component-library-react';

# ontario-badge

Use badges to indicate the status of a page or section.

Do not use a badge to bring a userâ€™s attention to important content.

## Usage guidance

Please refer to the [Ontario Design System](https://designsystem.ontario.ca/components/detail/badges.html) for current documentation guidance for badges.

## Configuration

Once the component package has been installed (see Ontario Design System Component Library for installation instructions), the badge component can be added directly into the project's code, and can be customized by updating the properties outlined [here](#properties). Please see the [examples](#examples) below for how to configure the component.

## Examples

Example of a badge component, where the user is explicitly passing in content through the `label` property.

```html
<ontario-badge label="Not started"></ontario-badge>
```

<div>
	<OntarioBadge
		label="Not started"
	></OntarioBadge>
</div>

This is another example of a badge. This time, the content is passed as a child of the `ontario-badge` component. A `colour` option is also passed.

```html
<ontario-badge colour="light-teal">Completed</ontario-badge>
```

<div>
	<OntarioBadge
		colour="light-teal"
	>
		Completed
	</OntarioBadge>
</div>

<!-- Auto Generated Below -->

## Properties

| Property        | Attribute         | Description                                                                                                               | Type                                                                          | Default     |
| --------------- | ----------------- | ------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------- | ----------- |
| `ariaLabelText` | `aria-label-text` | An aria label for screen readers. Used to provide more context to screen readers if necessary. This property is optional. | `string \| undefined`                                                         | `undefined` |
| `colour`        | `colour`          | The colour of the badge.                                                                                                  | `"black" \| "green" \| "grey" \| "light-teal" \| "red" \| "teal" \| "yellow"` | `'teal'`    |
| `label`         | `label`           | The label for the badge. Offical guidance is to keep the label length within 15 characters.                               | `string`                                                                      | `undefined` |

## Dependencies

### Used by

- [ontario-task](../ontario-task)

### Graph

```mermaid
graph TD;
  ontario-task --> ontario-badge
  style ontario-badge fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
