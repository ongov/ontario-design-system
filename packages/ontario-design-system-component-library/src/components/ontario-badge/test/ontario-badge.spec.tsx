import { newSpecPage } from '@stencil/core/testing';
import { OntarioBadge } from '../ontario-badge';

describe('ontario-badge', () => {
	it('should render the expected base html', async () => {
		const page = await newSpecPage({
			components: [OntarioBadge],
			html: `<ontario-badge>In progress</ontario-badge>`,
		});

		expect(page.root).toMatchSnapshot();
	});

	it('should override the host textContent with label prop output on the span element', async () => {
		const page = await newSpecPage({
			components: [OntarioBadge],
			html: `<ontario-badge label="completed">In progress</ontario-badge>`,
		});

		expect(page.root).toMatchSnapshot();
	});

	it('should render with the ontario-badge--default-light class on the span element', async () => {
		const page = await newSpecPage({
			components: [OntarioBadge],
			html: `<ontario-badge colour="lightTeal">In progress</ontario-badge>`,
		});

		expect(page.root).toMatchSnapshot();
	});

	it('should render the aria-label attribute on the span element', async () => {
		const page = await newSpecPage({
			components: [OntarioBadge],
			html: `<ontario-badge aria-label-text="This course is currently in progress.">In progress</ontario-badge>`,
		});

		expect(page.root).toMatchSnapshot();
	});
});
