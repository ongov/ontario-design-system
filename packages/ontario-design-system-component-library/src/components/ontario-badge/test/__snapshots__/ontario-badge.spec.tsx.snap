// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-badge should override the host textContent with label prop output on the span element 1`] = `
<ontario-badge label="completed">
  <template shadowrootmode="open">
    <span class="ontario-badge ontario-badge--default-heavy">
      completed
    </span>
  </template>
  In progress
</ontario-badge>
`;

exports[`ontario-badge should render the aria-label attribute on the span element 1`] = `
<ontario-badge aria-label-text="This course is currently in progress.">
  <template shadowrootmode="open">
    <span aria-label="This course is currently in progress." class="ontario-badge ontario-badge--default-heavy">
      In progress
    </span>
  </template>
  In progress
</ontario-badge>
`;

exports[`ontario-badge should render the expected base html 1`] = `
<ontario-badge>
  <template shadowrootmode="open">
    <span class="ontario-badge ontario-badge--default-heavy">
      In progress
    </span>
  </template>
  In progress
</ontario-badge>
`;

exports[`ontario-badge should render with the ontario-badge--default-light class on the span element 1`] = `
<ontario-badge colour="light-teal">
  <template shadowrootmode="open">
    <span class="ontario-badge ontario-badge--default-light">
      In progress
    </span>
  </template>
  In progress
</ontario-badge>
`;
